File: 3Sum_4Sum/3sumSmaller.py
def three_sum_smaller(nums, target):
    if len(nums) < 3:
        return 0

    nums.sort()
    count = 0

    for i in range(len(nums) - 2):
        second = i + 1
        third = len(nums) - 1

        while second < third:
            sum = nums[i] + nums[second] + nums[third]
            if sum < target:
                count += (third - second)
                second += 1
            else:
                third -= 1

    return count

nums = [-2, 0, 1, 3]
target = 2

print(three_sum_smaller(nums, target))

File: 3Sum_4Sum/3sumSmaller.rs
fn three_sum_smaller(nums: &mut Vec<i32>, target: i32) -> i32 {
    if nums.len() < 3 {
        return 0;
    }

    let mut count = 0;
    nums.sort();

    for i in 0..nums.len() - 2 {
        let mut second = i + 1;
        let mut third = nums.len() - 1;
        while second < third {
            let sum = nums[i] + nums[second] + nums[third];
            if sum < target {
                count += (third - second) as i32;
                second += 1;
            } else {
                third -= 1;
            }
        }
    }
    count
}

fn main() {
    let mut nums = vec![-2, 0, 1, 3];
    let target = 2;

    println!("{}", three_sum_smaller(&mut nums, target));
}
